lib/ukalloc/alloc.c

```diff
@@ -64,10 +68,18 @@ int uk_alloc_register(struct uk_alloc *a)
 	return 0;
 }
 
+#ifdef CONFIG_ARM64_FEAT_MTE
+struct metadata_ifpages {
+	__sz		size;
+	unsigned long	num_pages;
```

Comment:
Yes exactly, the idea was to tag only the memory requested by the user, up to `MTE_TAG_GRANULE_SIZE`. Both for performance and saving tag memory. I also imagine that storing the intended size could be useful for other purposes like stats etc. Do you think we should tag to page granularity instead?

Reply:
No, makes absolutely sense to restrict tagging to the actual allocation size (from a performance and security perspective). My intention was to make the number of places with ARCH-specific `#ifdefs` as small as possible to keep the code clean and simple. When replacing the `num_pages` field with the `size` field, we could use this with and without MTE. We only need to replace `num_pages` in the code with something like `ALIGN_UP(size, __PAGE_SIZE)`.

Reply:
@marcrittinghaus the problem I see with that is that we lose information required for `free()`. On the one hand, for the requirements of tagging we need to preserve the user size padded to MTE tag granule (to re-tag on `free()`). On the other hand, for the requirements of the allocator, we need to preserve the allocated pages. However during free this cannot always be obtained by aligning up `size + sizeof(struct metadata_ifpages)` to page size, because of the additional alignment potentially imposed if the user previously called `memalign_ifpages()`.

One idea to reduce the `#ifdef` noise a little bit is to unconditionally add the element `size` to `struct metadata_ifpages`, and also create a definition for the alignment, eg:
```
#ifdef CONFIG_ARM64_FEAT_MTE    
#define __align_metadata_ifpages __align(MTE_TAG_GRANULE);    
#else    
#define __align_metadata_ifpages    
#endif /* CONFIG_ARM64_FEAT_MTE */    

...
         
struct metadata_ifpages {    
        __sz            size;      /* user size */     
        unsigned long   num_pages; /* alloc pages */    
        void            *base;    
} __align_metadat_ifpages;
```
This will eliminate some noise like:
```
#ifdef CONFIG_ARM64_FEAT_MTE
        __sz size;
#endif /* CONFIG_ARM64_FEAT_MTE */
...
#ifdef CONFIG_ARM64_FEAT_MTE
        metadata->size = size;
#endif
```
but of course others will remain. What do you think?

Reply:
Ah, you are right. Yes, sounds good. Thanks for looking into this 😃